stdlib/public/core/Random.swift

```diff
@@ -130,20 +130,22 @@ extension RandomNumberGenerator {
 ///     let y = Int.random(in: 1...100, using: &g)
 ///
 /// `SystemRandomNumberGenerator` is automatically seeded, is safe to use in
-/// multiple threads, and uses a cryptographically secure algorithm whenever
-/// possible.
+/// multiple threads, and uses a cryptographically secure algorithm on all
+/// platforms where it is implemented.
```

Comment:
I'm not sure how we can make this claim, given that Swift can be ported to other platforms without notifying the main project or updating this documentation comment. I agree that we can strengthen the statement about the cryptographic quality, but would prefer staying on the line of stating which platforms provide that guarantee, rather than making a statement about "all platforms."

Reply:
I respectfully think I disagree. A ported version of Swift can choose to do anything, including things that wildly violate any of the documentation in this project: for example, it could implement Int64 as a 32-bit integer or, more subtly, a UInt64 as a 128-bit integer leading to surprising overflow divergences. We choose to not worry about that because if you do that, you're on your own lookout.

IMO we should restrict ourselves to documenting what our distribution of Swift does, and what any ported version of Swift should do. Any divergence from those documented rules will cause misbehaviour, but if we are requiring ourselves to document for that we will need to update every document with authoritative language to say "this does X unless it doesn't".

Reply:
Documentation is essentially our agreed-upon source of what Swift implementations actually do, so it doesn't make sense for a ported version of Swift to choose to do things that violate documented requirements (e.g. implementing `Int64` with 32 bits). In this case, though, the question of whether the default RNG is required to be cryptographically secure was discussed during the evolution process, and it is explicitly described as platform-dependent in the proposal itself:

> Each platform vendor will have the freedom to decide the specific implementation for this RNG for their platform. The standard library should document what specific RNG implementation is used on a specific platform. The aspiration is that this RNG should be cryptographically secure, provide reasonable performance, and should be thread safe. If a vendor is unable to provide these goals, they should document it clearly.

The documentation in this case should maintain that deference to the platform that Swift is running on, since whether or not the RNG is cryptographically secure really is up to the platform.

Reply:
Hm, in that case I think we should consider updating the proposal to either tighten the requirements on `SystemRandomNumberGenerator` or to propose a new one.

The problem with an API that says that it "should be" cryptographically secure is that it blesses the possibility that the API may not be. APIs that may not be cryptographically secure need to be treated as though they are not, except in cases where you can confidently confirm you only use it on platforms that provide a cryptographically secure implementation. That essentially requires guarding all uses of `SystemRandomNumberGenerator` in cryptographic contexts with `#if(os)` and `#if(swift)`.

APIs that can silently be non-secure are profoundly dangerous, and unless we resolve this I think we should be clearly warning users not to use `SystemRandomNumberGenerator` in libraries for cryptographic purposes.

Reply:
Let's move this to [the forums](https://forums.swift.org/t/clarify-the-cryptographic-properties-of-systemrandomnumbergenerator/27249).